buildsystem/coverage_report.gradle

```diff
@@ -42,24 +90,136 @@ if (variants != null) {
-        def kotlinTree = fileTree(dir: "${buildDir}/tmp/kotlin-classes/${variant.name}", excludes: fileFilter)
-        def mainSrc = "${project.projectDir}/src/main/java"
-        task("create${variantName}TestCoverage", type: JacocoReport, dependsOn: ["test${variantName}UnitTest", "connected${variantName}AndroidTest"]) {
-            group = "Reporting"
-            description = "Generate Jacoco coverage reports after running UI and Unit tests."
-            reports {
-                csv.enabled = false
-                xml.enabled = true
-                html.enabled = true
+        def debugTree = fileTree(
+                dir: "$buildDir/tmp/kotlin-classes/debug",
+                excludes: fileFilter
+        )
+        def mainSrc = "$projectDir/src/main/java"
+        sourceDirectories.from = files([mainSrc])
+        classDirectories.from = files([debugTree])
+        executionData.from = fileTree(
+                dir: project.buildDir,
+                includes: [
+                        'jacoco/testDebugUnitTest.exec',
+                        'outputs/code-coverage/connected/*coverage.ec'
+                ]
+        )
+    }
+}
+
+def setupAndroidCoverageVerification(threshold) {
+    task jacocoTestCoverageVerification(
+            type: JacocoCoverageVerification,
+            dependsOn: [
+                    'testDebugUnitTest',
+                    'testExistenceValidation'
+            ]
+    ) {
+        violationRules {
+            rule {
+                limit {
+                    minimum = threshold
+                }
+            }
+        }
+        final def coverageSourceDirs = [
+                "$projectDir/src/main/java",
+                "$projectDir/src/main/kotlin"
+        ]
+        final def debugTree = fileTree(
+                dir: "$buildDir/tmp/kotlin-classes/debug",
+                excludes: jacocoFileFilter
+        )
+        sourceDirectories.from = files(coverageSourceDirs)
+        classDirectories.from = files([debugTree])
+        executionData.from = fileTree(
+                dir: project.buildDir,
+                includes: ['jacoco/testDebugUnitTest.exec']
+        )
+    }
+}
+
+def setupKotlinCoverageVerification(threshold) {
+    jacocoTestCoverageVerification {
+        dependsOn test
+        dependsOn testExistenceValidation
+
+        violationRules {
+            rule {
+                limit {
+                    minimum = threshold
+                }
             }
+        }
+    }
+}
+
+def setupTestExistenceValidationTask() {
+    task testExistenceValidation(type: TestExistenceValidation)
+}
+
+project.ext {
+    jacocoCoverageThresholdDefault = 0.60
+
+    jacocoIgnoreList = [] // add module names to ignore
+    // Exclude file by names, packages or types. Such files will be ignored during test coverage
+    // calculation
+    jacocoFileFilter = [
+            '**/*App.*',
+            '**/*Application.*',
+            '**/*Activity.*',
+            '**/*Fragment.*',
```

Comment:
I am not sure if I agree with this. It is nice that the number then is only based on the coverage, but requirement that only logic code is tested is a compromise in the beginning. In the long run all code written by the team should be tested (including UI code) and it should be visible on codecov for which code this is not the case yet.

Reply:
We have two lists. One for doing coverage and one for failing the build if a threshold s not met. I will change the names of the filters to make this more clear. I think it is ok to exclude Fragments and Activities for the threshold calculation.

Reply:
Ahh yes, for the threshold go for it, totally makes sense.